android/library/src/test/java/io/logto/android/auth/browser/BrowserSignInFlowTest.kt

```diff
@@ -0,0 +1,162 @@
+import io.logto.android.exception.LogtoException
+import io.logto.android.model.OidcConfiguration
+import io.logto.android.model.TokenSet
+import io.logto.android.utils.Utils
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+import kotlinx.coroutines.newSingleThreadContext
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.setMain
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mock
+import org.mockito.Mockito.`when`
+import org.mockito.Mockito.spy
+import org.mockito.MockitoAnnotations
+import org.mockito.kotlin.anyOrNull
+import org.mockito.kotlin.argumentCaptor
+import org.mockito.kotlin.doAnswer
+import org.mockito.kotlin.eq
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.verify
+import org.robolectric.Robolectric
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+@ObsoleteCoroutinesApi
+@ExperimentalCoroutinesApi
+class BrowserSignInFlowTest {
+    private val mainThreadSurrogate = newSingleThreadContext("UI thread")
+
+    @Mock
+    private lateinit var logtoConfig: LogtoConfig
+
+    @Mock
+    private lateinit var logtoApiClient: LogtoApiClient
+
+    @Mock
+    private lateinit var onComplete: (exception: LogtoException?, tokenSet: TokenSet?) -> Unit
+
+    private val logtoExceptionCaptor = argumentCaptor<LogtoException>()
+    private val tokenSetCaptor = argumentCaptor<TokenSet>()
+
+    @Before
+    fun setUp() {
+        Dispatchers.setMain(mainThreadSurrogate)
+        MockitoAnnotations.openMocks(this)
+        `when`(logtoConfig.redirectUri).thenReturn("redirectUri")
+    }
+
+    @After
+    fun tearDown() {
+        Dispatchers.resetMain()
+        mainThreadSurrogate.close()
+    }
+
+    @Test
+    @Suppress("UNCHECKED_CAST")
+    fun shouldStartSignIn() {
+        val oidcConfiguration: OidcConfiguration = mock()
+        `when`(oidcConfiguration.authorizationEndpoint).thenReturn("authorizationEndpoint")
+
+        doAnswer {
+            val block = it.arguments[0] as (OidcConfiguration) -> Unit
+            block(oidcConfiguration)
+            null
+        }.`when`(logtoApiClient).discover(anyOrNull())
+
+        val activity: Activity = spy(Robolectric.buildActivity(Activity::class.java).get())
+
+        val browserSignInFlow = BrowserSignInFlow(
+            logtoConfig,
+            logtoApiClient,
+            onComplete
+        )
+
+        browserSignInFlow.start(activity)
+        verify(activity).startActivity(anyOrNull())
+    }
+
+    @Test
+    fun onResultShouldCompleteWithLogtoExceptionWithInvalidUri() {
+        val browserSignInFlow = BrowserSignInFlow(
+            logtoConfig,
+            logtoApiClient,
+            onComplete,
+        )
+
+        val invalidUri: Uri = mock()
+        `when`(invalidUri.getQueryParameter(eq(QueryKey.CODE))).thenReturn(null)
```

Comment:
similarly we could split   returnUri contains `ERROR` or `ERROR_DESCRIPTION` and returnUri `CODE` is null as two different exception case.